theano/sandbox/cuda/cusolver.py

```diff
@@ -0,0 +1,92 @@
+                CudaNdarrayType(broadcastable=[False])(),
+                CudaNdarrayType(broadcastable=[False])(),
+                CudaNdarrayType(broadcastable=[])()]
+
+    def make_node(self, inp):
+        inp = as_cuda_ndarray_variable(inp)
+        assert inp.ndim == 2
+
+        return theano.Apply(self, [inp], self.output_type(inp))
+
+    def make_thunk(self, node, storage_map, _, no_recycling=[]):
+        inputs = [storage_map[v] for v in node.inputs]
+        outputs = [storage_map[v] for v in node.outputs]
+
+        def thunk():
+            A = inputs[0][0]
+            m, n = A.shape
+            # LAPACK routines expect Fortran order
+            A_T = dimshuffle(A, (1, 0)).reshape((m, n))
+            A_copy = A_T.copy()
+
+            lda = max(1, m)
+            TAU = CudaNdarray.zeros((min(m, n),))
+            handle = cusolverDnCreate()
+            Lwork = cusolverDnSgeqrf_bufferSize(handle, m, n, A_copy.gpudata,
+                                                lda)
+            Work = CudaNdarray.zeros((max(1, Lwork),))
+            devInfo = CudaNdarray.zeros((1,))
+            cusolverDnSgeqrf(handle, m, n, A_copy.gpudata, lda, TAU.gpudata,
+                             Work.gpudata, Lwork, devInfo.gpudata)
+            cusolverDnDestroy(handle)
+            outputs[0][0] = reshape(A_copy, (n, m)).dimshuffle((1, 0))
+            outputs[1][0] = TAU
+            outputs[2][0] = Work
+            outputs[3][0] = devInfo
+
+        return thunk
+
+
+geqrf = GpuGeqrf()
+
+
+_libcusolver.cusolverDnSgeqrf.restype = int
+_libcusolver.cusolverDnSgeqrf.argtypes = [ctypes.c_void_p,
+                                          ctypes.c_int,
+                                          ctypes.c_int,
+                                          ctypes.c_void_p,
+                                          ctypes.c_int,
+                                          ctypes.c_void_p,
+                                          ctypes.c_void_p,
+                                          ctypes.c_int,
+                                          ctypes.c_void_p]
+
+
+def cusolverDnSgeqrf(handle, m, n, A, lda, TAU, Work, Lwork, devInfo):
+    status = _libcusolver.cusolverDnSgeqrf(handle, m, n, int(A), lda,
+                                           int(TAU), int(Work),
+                                           Lwork, int(devInfo))
+    cusolverCheckStatus(status)
+
+
+_libcusolver.cusolverDnSgeqrf_bufferSize.restype = int
+_libcusolver.cusolverDnSgeqrf_bufferSize.argtypes = [ctypes.c_void_p,
+                                                     ctypes.c_int,
+                                                     ctypes.c_int,
+                                                     ctypes.c_void_p,
+                                                     ctypes.c_int,
+                                                     ctypes.c_void_p]
+
+
+def cusolverDnSgeqrf_bufferSize(handle, m, n, A, lda):
```

Comment:
It's mostly because this is the way it's done in scikits.cuda, and it makes
it clear that it's a wrapper of a cuSOLVER function.

On Wed, Jun 10, 2015 at 1:00 PM, Frédéric Bastien <notifications@github.com>
wrote:

> In theano/sandbox/cuda/cusolver.py
> <https://github.com/Theano/Theano/pull/3013#discussion_r32140074>:
>
> > +    status = _libcusolver.cusolverDnSgeqrf(handle, m, n, int(A), lda,
> > +                                           int(TAU), int(Work),
> > +                                           Lwork, int(devInfo))
> > +    cusolverCheckStatus(status)
> > +
> > +
> > +_libcusolver.cusolverDnSgeqrf_bufferSize.restype = int
> > +_libcusolver.cusolverDnSgeqrf_bufferSize.argtypes = [ctypes.c_void_p,
> > +                                                     ctypes.c_int,
> > +                                                     ctypes.c_int,
> > +                                                     ctypes.c_void_p,
> > +                                                     ctypes.c_int,
> > +                                                     ctypes.c_void_p]
> > +
> > +
> > +def cusolverDnSgeqrf_bufferSize(handle, m, n, A, lda):
>
> I agree, nested function have limitation. I was not refering to doing
> that. I was refering to taking the inside of this function and putting it
> directly where you call it. This remove a python function call and make the
> code more readable as it is all closer.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/3013/files#r32140074>.
>